### PR TITLE
STRWEB-4: Update webpack to version 5

### PR DESCRIPTION
**Pull request by mkuklis**
https://issues.folio.org/browse/STRWEB-4

There are two corresponding PRs: 

* stripes-cli https://github.com/folio-org/stripes-cli/pull/270
* eslint-config-stripes https://github.com/folio-org/eslint-config-stripes/pull/85  

These will have to be merged after this PR is merged.

### Notes:

The biggest struggle with the upgrade was related the way compilation hooks were changed. The hooks object became frozen in webpack 5 (https://webpack.js.org/blog/2020-10-10-webpack-5-release/#hook-object-frozen). The way stripes webpack plugins communicate and the `StripesConfigPlugin` had to be refactored to follow this new pattern.   

I noticed a nice performance boost from both `stripes build` and `stripes serve`. For example `platform-complete` builds in half of the time compared to webpack 4. The initial `stripes serve` (the first time `stripes serve` is executed in a given folder) takes similar time to what we saw in webpack 4 but the following `stripes serve` is about 5x faster.

The best way to try it out before this work is merged is to checkout both `stripes-webpack` (STRWEB-4 branch) and `stripes-cli` (STCLI-187 branch) into a workspace together with a single ui module and add `"webpack": "^5.58.1"` to `resolutions` to make sure the correct version is being used (this may not be required).

````js
"resolutions": {
  "webpack": "^5.58.1",
}
````

Webpack 5 should open the door to a few nice features:

* Better code splitting with `dependsOn` https://webpack.js.org/configuration/entry-context/#dependencies
* Better optimization options https://webpack.js.org/configuration/optimization
* Module federation (Micro frontends) @skomorokh pointed to us originally: https://webpack.js.org/concepts/module-federation

### CODE REVIEW

**Review by zburke — CHANGES_REQUESTED** [review]
Wow! So excited to see you just grabbed this sucker and went for it! 99% fine to me, but the login screen logo doesn’t load for me and I see 
```
<img alt="..." src="[object Module]">
```
in the source vs 
```
<img alt="..." src="/img/tenant-assets/opentown-libraries-logo.c96ff678691e1a345321b50941335d81.png">
```
from folio-snapshot. I see the same behavior whether in a workspace or platform build. Trying to compile a build to an output directory, I get 
```
Module not found: Error: Can't resolve 'postcss-loader' in '/Users/zburke/temp/platform-core'
```
@JohnC-80, do you recall if there were any webpack-5 specific notes in the post-css work (#31) you did a few weeks ago that now come into play?

These are the only two glitches I've found so far. Nice work!

Edit: additional config details:
```
$ yarn --version
1.22.11
$ node --version
v16.10.0
platform-core(snapshot)*
$ npm --version
7.24.0
platform-core(snapshot)*
$ uname -a
Darwin LP-C02G8208ML85 20.6.0 Darwin Kernel Version 20.6.0: Mon Aug 30 06:12:21 PDT 2021; root:xnu-7195.141.6~3/RELEASE_X86_64 x86_64
```

**Comment by mkuklis** [comment]
Thank you for your feedback @JohnC-80. You are exactly right! I had to update `postcss-loader`. The version we had there was not compatible with webpack 5 anymore.

There is also `css-loader` which can be updated too but we will first need to migrate to `asset modules` https://webpack.js.org/guides/asset-modules before we can update `css-loader`. The current version works with webpack 5 so I'm going to leave it for now.

@zburke I would appreciate if you could give it another try with the most recent changes.

Oh I also had to upgrade `eslint-config-stripes`: https://github.com/folio-org/eslint-config-stripes/pull/85

**Review by zburke — APPROVED** [review]
With `postcss` `v8` et al in place, images are now rendering correctly for me. Impossible to overstate how great this news is! LGTM! 🎉 🎉 🎉

**Comment by mkuklis** [comment]
Good fine @zburke! I wonder what that `form.restart` is :)
I will take a look at these css issues.

**Comment by mkuklis** [comment]
@zburke it looks like the issue with colors is related to https://github.com/postcss/postcss-color-function which has been deprecated. There is a PR there to update postcss version https://github.com/postcss/postcss-color-function/pull/60 but it hasn't been merged. I pointed to that fork in `stripes-webpack` and it looks like that solved the problem for now.

There is also a newer lib (by the same author) which changes `color` function to `color-mod`: https://github.com/csstools/postcss-color-mod-function which is maintained and up to date but it looks like `color-mod` is also not part of the css spec anymore.

If we want to move to https://github.com/csstools/postcss-color-mod-function `stripes-components` would have to be updated to use `color-mod()` instead of `color()`.

There are couple places where `color()` is currently used in stripes-components:

````
./lib/Card/Card.css
./lib/variables.css
./lib/Datepicker/Calendar.css
./lib/Callout/Callout.css
./lib/Button/Button.css
./lib/sharedStyles/form.css
````

I think since this was removed from the css spec perhaps a better alternative would be to refactor these to not use `color()` or `color-mod()` at all. I will let @JohnC-80 decide what to do here.

For now I'm leaving `postcss-color-function` (from the fork) so we can move forward with the webpack upgrade.

@zburke thanks again for catching this!

File: package.json [hunk 1]
```diff
@@ -27,56 +27,58 @@
...
-    "postcss-custom-media": "^7.0.0",
-    "postcss-custom-properties": "^10.0.0",
-    "postcss-import": "^12.0.1",
-    "postcss-loader": "^4.0.0",
-    "postcss-media-minmax": "^4.0.0",
-    "postcss-nesting": "^7.0.1",
-    "postcss-url": "^8.0.0",
+    "postcss": "^8.3.9",
+    "postcss-calc": "^8.0.0",
+    "postcss-color-function": "onigoetz/postcss-color-function",
```
[the remark this answers is not in the document]
Comment (mkuklis) [reply package.json:55]: But I agree. I think the best would be to take care of the `color` right in `stripes-components`. Then we could remove `postcss-color-function` completely.

**Review by ryandberger — APPROVED** [review]

**Pull request merged**